File: ocaml/lib/kan/left_adjoint.ml
module LeftAdjoint (G : Functor.FUNCTOR) (H : Functor.FUNCTOR) = struct
  module GA = Functor.Apply(G)
  module HA = Functor.Apply(H)

  type 'a t = forall r. G.t (r -> 'a) -> H.t r

  let fmap f (c : 'a t) : 'b t =
    fun g -> H.fmap (fun x -> f (x g)) (c (G.fmap (fun f x -> f x) g))

  let liftF2 f (c1 : 'a t) (c2 : 'b t) : ('a -> 'b -> 'c) t =
    fun g ->
      HA.liftA2 (fun f g x y -> f (g x y))
                (c1 (G.fmap (fun f x y -> f (x, y)) g))
                (c2 (G.fmap (fun f x y -> f (y, x)) g))

  let apply (c1 : ('b -> 'c) t) (c2 : 'b t) : 'c t =
    fun g ->
      HA.apply (c1 (G.fmap (fun f x -> f x) g))
               (c2 (G.fmap (fun x f -> f x) g))

  let liftA2 = liftF2

  let liftA = apply

  let pure a : 'a t = fun g -> H.fmap (fun f -> f a) (G.fmap (fun x _ -> x) g)
end

module Adjoint (F : Functor.FUNCTOR) (U : Functor.FUNCTOR) (Adj : Adjunction.ADJUNCTION with module F = F and module U = U) = struct
  module LA = LeftAdjoint(F)(Identity)

  let adjointToLeftAdjoint : U.t 'a -> 'a LA.t = fun ua ->
    LA.LeftAdjoint (fun g -> Identity.Functor (Adj.right_adjunct (U.fmap (fun f -> f ()) ua) g))

  let leftAdjointToAdjoint : 'a LA.t -> U.t 'a = fun c ->
    Adj.left_adjunct (fun g -> LA.LeftAdjoint c (F.fmap (fun f x -> f (x ())) g)) ()
end

File: haskell/lib/Kan/Left_adjoint.hs
{-# LANGUAGE RankNTypes #-}
{-# LANGUAGE TypeOperators #-}

module LeftAdjoint
  ( LeftAdjoint (LeftAdjoint, runLeftAdjoint)
  , fmap
  , liftA2
  , apply
  , pure
  ) where

import Control.Applicative (liftA2)
import qualified Control.Category as C
import Control.Monad ((>=>))
import Data.Functor.Identity (Identity (Identity))

newtype LeftAdjoint g h a = LeftAdjoint { runLeftAdjoint :: g ((->) a) -> h () }

instance (Functor g, Functor h) => Functor (LeftAdjoint g h) where
  fmap f (LeftAdjoint c) = LeftAdjoint $ \g -> fmap (fmap f) (c (fmap (fmap ($ ())) g))

instance (Applicative g, Applicative h) => Applicative (LeftAdjoint g h) where
  pure a = LeftAdjoint $ \g -> pure (const a) <*> g
  LeftAdjoint f <*> LeftAdjoint x = LeftAdjoint $ \g -> liftA2 (liftA2 ($)) (f (fmap (fmap (,)) g)) (x (fmap (fmap snd) g))

liftA2 :: (Applicative g, Applicative h) => (a -> b -> c) -> LeftAdjoint g h a -> LeftAdjoint g h b -> LeftAdjoint g h c
liftA2 f (LeftAdjoint c1) (LeftAdjoint c2) = LeftAdjoint $ \g ->
  liftA2 (liftA2 ($)) (c1 (fmap (fmap (,)) g)) (c2 (fmap (fmap snd) g)) >>= \f' x' ->
    pure (f' x')

apply :: (Applicative g, Applicative h) => LeftAdjoint g h (a -> b) -> LeftAdjoint g h a -> LeftAdjoint g h b
apply = liftA2 ($)

data Adjunction f u = Adjunction
  { leftAdjunct :: forall x. (f x -> u x) -> u x
  , rightAdjunct :: forall x. u x -> f x
  }

adjointToLeftAdjoint :: (Adjunction f u) => u a -> LeftAdjoint f Identity a
adjointToLeftAdjoint ua = LeftAdjoint $ \g -> Identity $ rightAdjunct (fmap ($()) ua) g

leftAdjointToAdjoint :: (Adjunction f u) => LeftAdjoint f Identity a -> u a
leftAdjointToAdjoint (LeftAdjoint c) = leftAdjunct $ \g -> runIdentity $ c (Identity . g)